lib/class-wp-block-supports.php

```diff
@@ -0,0 +1,454 @@
+		$instance = self::get_instance();
+		$instance->register_attributes();
+		add_action( 'wp_footer', array( $instance, 'enqueue_styles' ) );
+		add_filter( 'render_block', array( $instance, 'apply_block_supports' ), 10, 2 );
+	}
+
+	/**
+	 * Enqueues an empty style to which inline styles may be attached.
+	 */
+	public function enqueue_styles() {
+		wp_enqueue_style( 'wp-block-supports' );
+	}
+
+	/**
+	 * Generates an array of HTML attributes, such as classes, by applying to
+	 * the given block all of the features that the block supports.
+	 *
+	 * @param  array $parsed_block Block as parsed from content.
+	 * @return array               Array of HTML attributes.
+	 */
+	public function get_new_block_attributes( $parsed_block ) {
+		if (
+			empty( $parsed_block ) ||
+			! isset( $parsed_block['attrs'] ) ||
+			! isset( $parsed_block['blockName'] )
+		) {
+			// TODO: handle as error?
+			return array();
+		}
+
+		$block_attributes = $parsed_block['attrs'];
+		$block_type       = WP_Block_Type_Registry::get_instance()->get_registered(
+			$parsed_block['blockName']
+		);
+
+		if ( empty( $block_type ) ) {
+			// TODO: handle as error?
+			return array();
+		}
+
+		$output = array();
+		foreach ( $this->config as $feature_name => $feature_config ) {
+			if (
+				! $this->get_block_support(
+					$block_type->supports,
+					$feature_name,
+					$feature_config['default']
+				)
+			) {
+				continue;
+			}
+			$new_attributes = call_user_func(
+				$feature_config['callback'],
+				// Pick from $block_attributes according to feature attributes.
+				array_intersect_key(
+					$block_attributes,
+					array_flip( array_keys( $feature_config['attributes'] ) )
+				),
+				$block_type->name
+			);
+			if ( ! empty( $new_attributes ) ) {
+				foreach ( $new_attributes as $attribute_name => $attribute_value ) {
+					if ( empty( $output[ $attribute_name ] ) ) {
+						$output[ $attribute_name ] = '';
+					}
+					$output[ $attribute_name ] .= " $attribute_value";
+				}
+			}
+		}
+		return $output;
+	}
+
+	/**
+	 * TODO.
+	 *
+	 * @param  string $block_content Rendered block content.
+	 * @param  array  $block         Block object.
+	 * @return string                Transformed block content.
+	 */
+	public function apply_block_supports( $block_content, $block ) {
+		if ( ! isset( $block['attrs'] ) ) {
+			return $block_content;
+		}
+
+		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+		// If no render_callback, assume styles have been previously handled.
+		if ( ! $block_type || ! $block_type->render_callback ) {
+			return $block_content;
+		}
+
+		$new_attributes = $this->get_new_block_attributes( $block );
+		if ( ! empty( $new_attributes ) ) {
+			return $this->inject_attributes( $block_content, $new_attributes );
+		}
+
+		return $block_content;
+	}
+
+	/**
+	 * Constructs the configuration for all supported block-supports features.
+	 */
+	private function load_config() {
```

Comment:
I don't really like this giant config here, for me each one of these should be its own file. it doesn't have to be a public API (can use a private one if the class is private) to be the case.